### ISSUE TITLE
CWWKE0702E: Could not resolve module: com.ibm.ws.ejbcontainer.remote [852] Bundle was not resolved because of a uses constraint violation.

### ISSUE THREAD

**Issue by tpridmore** [OpenLiberty/open-liberty#20082]
Description
We are seeing the error below on startup of WebSphere Liberty 22.0.0.1 for a couple of our server configurations. The message looks very similar to issue #14936 which affected our product last year.  For that issue we added osgi.checkConfiguration=false to bootstrap.properties as a work-around and Liberty started fine. The work-around was removed sometime after integrating WebSphere Liberty 21.0.0.3 and all worked well until recently.  I have implemented the work-around again to mitigate the current occurrence.

```
Adding osgi.checkConfiguration=false to bootstrap.properties as a work-around also works for the current issue. Is there an open ticket for the problem?
[2/7/22 14:28:18:665 UTC] 00000028 remote E CWWKE0702E: Could not resolve module: com.ibm.ws.ejbcontainer.remote [852]
Bundle was not resolved because of a uses constraint violation.
org.apache.felix.resolver.reason.ReasonException: Uses constraint violation. Unable to resolve resource com.ibm.ws.ejbcontainer.remote [osgi.identity; type="osgi.bundle"; version:Version="1.0.60.cl220120220103-1900"; osgi.identity="com.ibm.ws.ejbcontainer.remote"] because it is exposed to package 'javax.rmi.CORBA' from resources com.ibm.ws.org.apache.yoko.rmi.spec.1.5 [osgi.identity; type="osgi.bundle"; version:Version="1.0.60.cl220120220103-1900"; osgi.identity="com.ibm.ws.org.apache.yoko.rmi.spec.1.5"] and org.eclipse.osgi [osgi.identity; osgi.identity="org.eclipse.osgi"; type="osgi.bundle"; version:Version="3.16.0.cl220120220103-1900"; singleton:="true"] via two dependency chains.

Chain 1:
com.ibm.ws.ejbcontainer.remote [osgi.identity; type="osgi.bundle"; version:Version="1.0.60.cl220120220103-1900"; osgi.identity="com.ibm.ws.ejbcontainer.remote"]
import: (osgi.wiring.package=javax.rmi.CORBA)
|
export: osgi.wiring.package: javax.rmi.CORBA
com.ibm.ws.org.apache.yoko.rmi.spec.1.5 [osgi.identity; type="osgi.bundle"; version:Version="1.0.60.cl220120220103-1900"; osgi.identity="com.ibm.ws.org.apache.yoko.rmi.spec.1.5"]

Chain 2:
com.ibm.ws.ejbcontainer.remote [osgi.identity; type="osgi.bundle"; version:Version="1.0.60.cl220120220103-1900"; osgi.identity="com.ibm.ws.ejbcontainer.remote"]
import: (&(osgi.wiring.package=com.ibm.ejs.container)(&(version>=3.0.0)(!(version>=4.0.0))))
|
export: osgi.wiring.package=com.ibm.ejs.container; uses:=javax.rmi.CORBA
com.ibm.ws.ejbcontainer [osgi.identity; type="osgi.bundle"; version:Version="1.0.60.cl220120220103-1900"; osgi.identity="com.ibm.ws.ejbcontainer"]
import: (osgi.wiring.package=javax.rmi.CORBA)
|
export: osgi.wiring.package: javax.rmi.CORBA
org.eclipse.osgi [osgi.identity; osgi.identity="org.eclipse.osgi"; type="osgi.bundle"; version:Version="3.16.0.cl220120220103-1900"; singleton:="true"]
```

Configuration
Feature list for each of the two failing server configurations...
```
    <featureManager onError="FAIL">
        <feature>adminCenter-1.0</feature>
        <feature>webProfile-8.0</feature>
        <feature>mpHealth-2.2</feature>
        <feature>mpConfig-1.4</feature>
        <feature>appSecurity-2.0</feature>
        <feature>transportSecurity-1.0</feature>
        <feature>jdbc-4.2</feature>
        <feature>json-1.0</feature>
        <feature>jsp-2.3</feature>
        <feature>mdb-3.2</feature>
        <feature>localConnector-1.0</feature>
        <feature>eventLogging-1.0</feature>
        <feature>restConnector-2.0</feature>
        <feature>wmqJmsClient-2.0</feature>
        <feature>ejb-3.2</feature>
        <feature>javaMail-1.6</feature>
        <feature>concurrent-1.0</feature>
        <feature>websocket-1.1</feature>
        <feature>batchManagement-1.0</feature>
        <feature>jaxws-2.2</feature>
        <feature>ldapRegistry-3.0</feature>
        <feature>concurrent-1.0</feature>
    </featureManager>
```
    
```
    <featureManager onError="FAIL">
        <feature>adminCenter-1.0</feature>
        <feature>appSecurity-2.0</feature>
        <feature>ejb-3.2</feature>
        <feature>eventLogging-1.0</feature>
        <feature>javaMail-1.6</feature>
        <feature>jaxws-2.2</feature>
        <feature>jdbc-4.2</feature>
        <feature>json-1.0</feature>
        <feature>jsp-2.3</feature>
        <feature>localConnector-1.0</feature>
        <feature>mdb-3.2</feature>
        <feature>mpHealth-2.2</feature>
        <feature>mpConfig-1.4</feature>
        <feature>restConnector-2.0</feature>
        <feature>transportSecurity-1.0</feature>
        <feature>webProfile-8.0</feature>
        <feature>wmqJmsClient-2.0</feature>
        <feature>ejbRemote-3.2</feature> 
        <feature>ldapRegistry-3.0</feature>
    </featureManager>
```

Environment
```
product = WebSphere Application Server 22.0.0.1 (wlp-1.0.60.cl220120220103-1900)
wlp.install.dir = /opt/ibm/wlp/
server.output.dir = /opt/ibm/wlp/output/defaultServer/
java.home = /opt/ibm/java/jre
java.version = 1.8.0_311
java.runtime = Java(TM) SE Runtime Environment (8.0.7.0 - pxa6480sr7-20211025_01(SR7))
os = Linux (4.18.0-305.28.1.el8_4.x86_64; amd64) (en_US)
process = 175@10.254.15.38
```

This is a container deployment...
Red Hat Enterprise Linux 8.5 (Ootpa)

**Comment by tpridmore** [comment]
`osgi.checkConfiguration=false` is no longer working.  This is now an urgent issue.

**Comment by tjwatson** [comment]
I have tried to reproduce with the two following `Dockerfile` contents but have not been able to generate the error:

```
FROM icr.io/appcafe/websphere-liberty:22.0.0.1-kernel-java8-ibmjava-ubi

RUN echo '<?xml version="1.0" encoding="UTF-8"?>\
<server description="new server">\
    <!-- Enable features -->\
    <featureManager onError="FAIL">\
        <feature>adminCenter-1.0</feature>\
        <feature>webProfile-8.0</feature>\
        <feature>mpHealth-2.2</feature>\
        <feature>mpConfig-1.4</feature>\
        <feature>appSecurity-2.0</feature>\
        <feature>transportSecurity-1.0</feature>\
        <feature>jdbc-4.2</feature>\
        <feature>json-1.0</feature>\
        <feature>jsp-2.3</feature>\
        <feature>mdb-3.2</feature>\
        <feature>localConnector-1.0</feature>\
        <feature>eventLogging-1.0</feature>\
        <feature>restConnector-2.0</feature>\
        <feature>wmqJmsClient-2.0</feature>\
        <feature>ejb-3.2</feature>\
        <feature>javaMail-1.6</feature>\
        <feature>concurrent-1.0</feature>\
        <feature>websocket-1.1</feature>\
        <feature>batchManagement-1.0</feature>\
        <feature>jaxws-2.2</feature>\
        <feature>ldapRegistry-3.0</feature>\
        <feature>concurrent-1.0</feature>\
    </featureManager>\
</server>' > /config/server.xml

RUN configure.sh

```

```
FROM icr.io/appcafe/websphere-liberty:22.0.0.1-kernel-java8-ibmjava-ubi

RUN echo '<?xml version="1.0" encoding="UTF-8"?>\
<server description="new server">\
    <!-- Enable features -->\
    <featureManager onError="FAIL">\
        <feature>adminCenter-1.0</feature>\
        <feature>appSecurity-2.0</feature>\
        <feature>ejb-3.2</feature>\
        <feature>eventLogging-1.0</feature>\
        <feature>javaMail-1.6</feature>\
        <feature>jaxws-2.2</feature>\
        <feature>jdbc-4.2</feature>\
        <feature>json-1.0</feature>\
        <feature>jsp-2.3</feature>\
        <feature>localConnector-1.0</feature>\
        <feature>mdb-3.2</feature>\
        <feature>mpHealth-2.2</feature>\
        <feature>mpConfig-1.4</feature>\
        <feature>restConnector-2.0</feature>\
        <feature>transportSecurity-1.0</feature>\
        <feature>webProfile-8.0</feature>\
        <feature>wmqJmsClient-2.0</feature>\
        <feature>ejbRemote-3.2</feature> \
        <feature>ldapRegistry-3.0</feature>\
    </featureManager>\
</server>' > /config/server.xml

RUN configure.sh


```

I suspect something could be going wrong during your image build with the `configure.sh` run.  That is where liberty will prime its cache of installed bundles and do the initial resolution.  That resolution rules and state are cached.  But this only makes sense if you observe a container that consistently fails and then if you rebuild the container it may start to consistently launch.

First off, do you use the `configure.sh` in your image build?  I may need to know some more details on how you build your image.

**Comment by tpridmore** [comment]
Yes, we do use `configure.sh` in the image builds.

If we build and launch successfully, then that image always works.  If we build and launch with error, then that image always fails.

**Comment by tjwatson** [comment]
I believe I have found an issue (and a fix) in Equinox that could be causing this (eclipse-equinox/equinox.framework#17).

This is similar to #14936 but the mechanism in the bootup of the framework is allow bundles to resolve without the region isolation in place.  This is allowing resolution to the package `javax.rmi.CORBA` from the JVM (system.bundle) incorrectly.  This ultimately causes the uses constraint violation.  Until we pull in the fixed Equinox (or patch the one we are using) we should disable the osgi.checkConfiguration again, like we did in #14940

In the long run we may want to consider always disabling that when deployed in containers.  That option is really only useful for development scenarios.
